### Simple Linear Regression Example 3.3:
For the **Advertising** data we can determine the standard errors of the estimated coefficients with the help of **Python**:

In [1]:
import pandas as pd
import statsmodels.api as sm

# Load data
df = pd.read_csv('./data/Advertising.csv')
x = df['TV']
y = df['sales']

# Linear Regression using statsmodels.api
x_sm = sm.add_constant(x)
model = sm.OLS(y, x_sm).fit()

# Now we can print a summary, 
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Fri, 25 Feb 2022   Prob (F-statistic):           1.47e-42
Time:                        17:56:43   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.0326      0.458     15.360      0.0

In the table under **coef**, the listed values are $\hat{\beta}_{0} $ and $ \hat{\beta}_{1} $. In the column **Std.\,Error** we find the values $ 0.457843 $  and $ 0.002691 $ for the two standard errors $ se (\hat{\beta}_{0}) $ and  $ se (\hat{\beta}_{1}) $. They correspond to the average deviations of the estimated values of $\hat{\beta}_{0} $ and $ \hat{\beta}_{1} $ which are $ 7.0326 $ and $ 0.04754 $.
In the right top part, we find some constants describing the fit, for example **R-squared** which is the portion of variance explained by the fit. 

### Simple Linear Regression Example 3.4
For the **Advertising** data, we want to compute the p-value of  $\beta_1$  of the least squares model for the regression of number of units sold on TV advertising budget.

The least squares estimate for $\beta_1$ is 0.047537 and the standard error of $\hat{\beta}_1$.

Hence, we find for the realization of the statistic $T$
\begin{equation*}
t
=\dfrac{\hat{\beta}_{1}-0}{se (\hat{\beta}_{1})}
=\dfrac{0.047537-0}{0.002691}
=17.66518
\end{equation*}


Furthermore, we find in the **Python**-output, that the number of degrees of freedom is $198$, which is listed under **Df Residuals**. The number of degrees of freedom is given by  $n-2$, it follows that in total there are $n=200$ data points. This number corresponds to the number of markets in the data set **Advertising**.

Since the coefficient for $\hat{\beta}_1$ is very large relative to its standard error, so the t-statistic is also large. The probability of seeing such a value if $H_{0}$ is true is virtually zero. Hence we can conclude that $\beta_{1}\neq 0$. 

We now will compute the p-value, that is, the probability of observing a value of the t-statistic larger than $|t|=17.66518$. Assuming $\beta_1=0$, $T$ will follow a t-distribution with  $n-2=198$ degrees of freedom. Then the (two-sided) p-value can be determined with the help of **Python**

In [5]:
from scipy.stats import t

p_two_sided = 2 * (1 - t.cdf(17.66518, 198))
print(p_two_sided)

0.0


Since the alternative hypothesis is two-sided, we need to multiply the one-sided p-value by two in order to obtain the two-sided p-value. 
The corresponding  p-value : $0.000$ is listed under **P(>|t|)**. Since it is zero, hence smaller than a siginificance level of $ 0.05 $ , we *reject* the null hypothesis  $\beta_{1}=0$ in favor of the alternative hypothesis $\beta_{1}\neq 0 $. We therefore find that there clearly is a relationship between **TV** and **sales**.

### Simple Linear Regression Example 3.5
In the case of the **Advertising** data, the 95% confidence interval can be found with the help of the **conf\_int**-function: 

In [6]:
import numpy as np
# Confidence interval found using conf_int method
print(np.round(model.conf_int(alpha=0.05), 4))

            0       1
const  6.1297  7.9355
TV     0.0422  0.0528


The 95%-confidence interval for $\beta_{0}$ thus is
\begin{equation*}
[6.130,7.935]
\end{equation*}
and the 95%-confidence interval for $\beta_{1}$
\begin{equation*}
[0.042,0.053]
\end{equation*}
Therefore, we can conclude that in the absence of any advertising, sales will, on average, fall somewhere between $ 6130 $ and $ 7935 $ units. Furthermore, for each  CHF 1000 increase in television advertising, there will be an average increase in sales of between 42 and 53 units. 

1. The exact formula for the 95% confidence interval for the regression coefficient $\beta_i$ is

\begin{equation*}
\left [
\hat{\beta}_i- t_{0.975;n-2}\cdot se (\hat{\beta}_{i}),\hat{\beta}_{i}+ t_{0.975;n-2}\cdot se (\hat{\beta}_{i})
\right ]
\end{equation*}

where $t_{0.975;n-1}$ is the 97.5% quantile of a t-distribution with  $n-2$ degrees of freedom. 

2. With **Python** we determine the  97.5% quantile of a t-distribution as follows

In [7]:
# the 97.5% quantile of a t-distribution:
q_975 = t.ppf(0.975, 18)
print(np.round(q_975, 4))

2.1009


which is approximately 2.